File: guests/swift/todos.swift
// The todos scene, Swift port — guests/rust/todos.rs,
// tools/scenes/todos.steps.

import Foundation

struct Todo: KayaGen {
    var title: String
    var done: Bool
}

let app = KayaApp()

var draft = ""

app.build { tx in
    // The two items are the whole undo surface an app declares
    // (docs/undo-plan.md D1-D6).
    let edit = tx.menu(
        "Edit",
        items: [
            tx.item("Undo", role: KayaAppTx.roleUndo),
            tx.item("Redo", role: KayaAppTx.roleRedo),
        ])
    // NO onUndone AND NO onRedone: this scene's whole state is the collection
    // and the signal derived from it, both of which the core restores.
    tx.window(title: "todos", menus: [edit])
    let todos = todoCollection(tx)
    let itemsLeft = todos.derive(tx) { items in
        let n = items.filter { !$0.value.done }.count
        return .str(n == 1 ? "1 item left" : "\(n) items left")
    }

    let root = tx.column {
        let field = tx.entry { _, text in draft = text }
        tx.button("Add") { tx in
            if draft.isEmpty { return }
            // The derive's write is in THIS batch, so the step's inverse
            // carries "0 items left".
            tx.undoable("add \(draft)")
            // The binding mints the key (docs/fresh-key-plan.md).
            todos.insertFresh(tx, Todo(title: draft, done: false))
            // `clear` inside a group is refused at apply (docs/undo-plan.md D4).
            app.post { tx in
                tx.clear(field)
                tx.focus(field)
            }
        }
        tx.label(bind: itemsLeft)
        for row in todos.rows {
            row.row {
                row.checkbox(row.done) { tx, keys, checked in
                    todos.patch(tx, keys[0]).set(\.done, checked)
                }
                row.label(row.title)
            }
        }
    }
    tx.mount(root)
}

app.run()

File: guests/ocaml/todos.ml
(* The todos scene from OCaml, on the let surface with the construction
   sugar: the record declaration is the schema ([@@deriving kaya_gen]).

   The app names no todo — the key comes from [insert_record_fresh]
   (docs/fresh-key-plan.md).

   THE ADD IS ONE STEP AND THIS FILE WRITES NO UNDO CODE FOR IT. The
   derive's write rides the same transaction, so the group banks both
   halves and the core restores them together.

   Build like milestone2.ml, then run with KAYA_SELFTEST=todos. *)

open Kaya_wire
open Kaya_app

type todo = { title : string; done_ : bool } [@@deriving kaya_gen]

let () =
  let app = Kaya_app.create () in

  (* The fold: widget-owned state arrives as occurrences, and the app's
     copy is this ref rather than a widget read. *)
  let draft = ref "" in

  build app (fun () ->
     let todos = collection_of todo_record in
     (* A derived signal: the binding recomputes it from the collection
        after every mutation, so no handler mentions it. *)
     let items_left =
       derive todos (fun entries ->
           let n = List.length (List.filter (fun (_, t) -> not t.done_) entries) in
           Str (if n = 1 then "1 item left" else Printf.sprintf "%d items left" n))
     in
     let field = entry ~on_change:(fun text -> draft := text) () in
     let on_add () =
       let d = !draft in
       if d = "" then ()
       else begin
         undoable (Printf.sprintf "add %s" d);
         ignore (insert_record_fresh todos { title = d; done_ = false });
         (* Emptying the form is NOT part of the step, so it needs its
            own transaction — and [clear] inside an undoable group is
            refused at apply anyway (docs/undo-plan.md D4). [post], not a
            nested [build]: a handler already is a transaction
            (tools/check-ambient-tx.py). *)
         post app (fun () ->
             clear field;
             focus field)
       end
     in
     let on_toggle keys checked =
       (* One field's delta: the title never travels. [todo_patch] is
          ppx-generated, one optional labelled argument per field. *)
       todo_patch ~done_:checked todos (List.hd keys)
     in

     window ~title:"todos"
       ~menus:
         [
           menu ~label:"Edit"
             [
               item ~label:"Undo" ~role:role_undo;
               item ~label:"Redo" ~role:role_redo;
             ];
         ]
       ();

     let root =
       column
         [
           w field;
           button ~text:"Add" ~on_click:on_add;
           label ~bind:items_left;
           each (record_handle todos) (fun () ->
               Tpl.(
                 row
                   [
                     checkbox ~bind_field:todo_done_ ~on_toggle;
                     label ~bind_field:todo_title;
                   ]
                   ()));
         ]
         ()
     in
     mount root);

  exit (run app)
